cli/lsp/repl.rs

```diff
@@ -0,0 +1,308 @@
+          })
+        })
+        .filter(|item| {
+          // filter the results to only exact matches
+          let text = &text_info.text_str()
+            [item.span.lo.0 as usize..item.span.hi.0 as usize];
+          item.new_text.starts_with(text)
+        })
+        .map(|mut item| {
+          // convert back to a line position
+          item.span = Span::new(
+            item.span.lo - before_line_len,
+            item.span.hi - before_line_len,
+            Default::default(),
+          );
+          item
+        })
+        .collect()
+    })
+    .await
+  }
+
+  async fn did_change(&mut self, new_text: &str) {
+    self.check_cwd_change().await;
+    let new_text = if new_text.ends_with('\n') {
+      new_text.to_string()
+    } else {
+      format!("{}\n", new_text)
+    };
+    self.document_version += 1;
+    let current_line_count =
+      self.document_text.chars().filter(|c| *c == '\n').count() as u32;
+    let pending_line_count =
+      self.pending_text.chars().filter(|c| *c == '\n').count() as u32;
+    self
+      .language_server
+      .did_change(DidChangeTextDocumentParams {
+        text_document: VersionedTextDocumentIdentifier {
+          uri: self.get_document_specifier(),
+          version: self.document_version,
+        },
+        content_changes: vec![TextDocumentContentChangeEvent {
+          range: Some(Range {
+            start: Position::new(current_line_count, 0),
+            end: Position::new(current_line_count + pending_line_count, 0),
+          }),
+          range_length: None,
+          text: new_text.to_string(),
+        }],
+      })
+      .await;
+    self.pending_text = new_text;
+  }
+
+  async fn check_cwd_change(&mut self) {
+    // handle if the cwd changes
+    let cwd_uri = get_cwd_uri().unwrap();
+    if self.cwd_uri != cwd_uri {
+      self
+        .language_server
+        .did_close(DidCloseTextDocumentParams {
+          text_document: TextDocumentIdentifier {
+            uri: self.get_document_specifier(),
+          },
+        })
+        .await;
+      self.cwd_uri = cwd_uri;
+      self.document_version = 0;
+      self.open_current_document().await;
+    }
+  }
+
+  async fn open_current_document(&self) {
+    self
+      .language_server
+      .did_open(DidOpenTextDocumentParams {
+        text_document: TextDocumentItem {
+          uri: self.get_document_specifier(),
+          language_id: "typescript".to_string(),
+          version: self.document_version,
+          text: format!("{}{}", self.document_text, self.pending_text),
+        },
+      })
+      .await;
+  }
+
+  fn get_document_specifier(&self) -> ModuleSpecifier {
+    self.cwd_uri.join("$deno$repl.ts").unwrap()
+  }
+}
+
+async fn with_logging_disabled<TReturn>(
+  future: impl Future<Output = TReturn>,
+) -> TReturn {
+  logger::set_logging_enabled(false);
+  let result = future.await;
+  // ensure this is set back even on error of a Result
+  logger::set_logging_enabled(true);
```

Comment:
Toggling globals like that feels somewhat unsafe. I suppose it works because this function is only called by `&mut self` methods but it's a bit brittle.